cli/cli_app.go

```diff
@@ -183,54 +189,165 @@ func runCommand(command string, terragruntOptions *options.TerragruntOptions) (f
+// the respective terraform command (e.g. 'terraform apply')
+//
+// The terragruntOptions are assumed to be the options for running the original terragrunt command.
+//
+// If terraformSource is specified, then arguments to download the terraform source will be appended to the init command.
+//
+// This method will return an error and NOT run terraform init if the user has disabled Auto-Init
+func runTerraformInit(terragruntOptions *options.TerragruntOptions, terragruntConfig *config.TerragruntConfig, terraformSource *TerraformSource) error {
+
+	// Prevent Auto-Init if the user has disabled it
+	if firstArg(terragruntOptions.TerraformCliArgs) != CMD_INIT && !terragruntOptions.AutoInit {
+		return errors.WithStackTrace(InitNeededButDisabled("Cannot continue because init is needed, but Auto-Init is disabled.  You must run 'terragrunt init' manually."))
+	}
+
+	// Need to clone the terragruntOptions, so the TerraformCliArgs can be configured to run the init command
+	initOptions := terragruntOptions.Clone(terragruntOptions.TerragruntConfigPath)
+	initOptions.TerraformCliArgs = []string{CMD_INIT}
+	initOptions.WorkingDir = terragruntOptions.WorkingDir
+
+	// Only add the arguments to download source if terraformSource was specified
+	downloadSource := terraformSource != nil
+	if downloadSource {
+		v0_10_0, err := version.NewVersion("v0.10.0")
+		if err != nil {
 			return err
 		}
-	}
 
-	// If the command is 'init', stop here. That's because ConfigureRemoteState above will have already called
-	// terraform init if it was necessary, and the below RunTerraformCommand would end up calling init without
-	// the correct remote state arguments, which is confusing.
-	if terragruntOptions.TerraformCliArgs[0] == CMD_INIT {
-		terragruntOptions.Logger.Println("Running 'init' manually is not necessary: Terragrunt will call it automatically when needed before running other Terraform commands")
-		return nil
+		if terragruntOptions.TerraformVersion.LessThan(v0_10_0) {
+			// Terraform versions < 0.10.0 specified the module source as an argument (rather than the -from-module option)
+			initOptions.AppendTerraformCliArgs(terraformSource.CanonicalSourceURL.String())
+		} else {
+			// Terraform versions >= 0.10.0 specify the module source using the -from-module option
+			initOptions.AppendTerraformCliArgs("-from-module=" + terraformSource.CanonicalSourceURL.String())
+		}
+		initOptions.AppendTerraformCliArgs(terraformSource.DownloadDir)
 	}
 
-	return shell.RunTerraformCommand(terragruntOptions, terragruntOptions.TerraformCliArgs...)
+	return runTerragruntWithConfig(initOptions, terragruntConfig, downloadSource)
+}
+
+// Returns an error if allowSourceDownload is false, and terragruntOptions.TerraformCliArgs contains source download related arguments
+func verifySourceDownloadArguments(allowSourceDownload bool, terragruntOptions *options.TerragruntOptions) error {
+	if !allowSourceDownload && len(terragruntOptions.TerraformCliArgs) > 1 {
```

Comment:
You can reduce nesting a bit by inverting the if-statement and returning `nil` immediately if it's true. Then the body of the if statement is moved outside of it and is just the regular method body.